examples/optsCommands.c

```diff
@@ -0,0 +1,161 @@
+		      keyNew (SPEC_BASE_KEY "/setter/keyname", KEY_META, "description", "name of the key to write", KEY_META, "args",
+			      "indexed", KEY_META, "args/index", "0", KEY_END),
+		      keyNew (SPEC_BASE_KEY "/setter/value", KEY_META, "description", "value to be written", KEY_META, "args", "indexed",
+			      KEY_META, "args/index", "1", KEY_END),
+		      keyNew (SPEC_BASE_KEY "/dynamic/#", KEY_META, "description", "dynamically call a user-supplied command", KEY_META,
+			      "args", "remaining", KEY_END),
+		      KS_END);
+}
+
+int main (int argc, const char ** argv)
+{
+	KeySet * ks = createSpec ();
+	Key * errorKey = keyNew (BASE_KEY, KEY_END);
+
+	int result = elektraGetOpts (ks, argc, argv, (const char **) environ, errorKey);
+	if (result == -1)
+	{
+		fprintf (stderr, "ERROR: %s\n", keyString (keyGetMeta (errorKey, "error/reason")));
+		keyDel (errorKey);
+		ksDel (ks);
+		return EXIT_FAILURE;
+	}
+
+	if (result == 1)
+	{
+		char * help = elektraGetOptsHelpMessage (errorKey, NULL, NULL);
+		fprintf (stderr, "%s\n", help);
+		elektraFree (help);
+		keyDel (errorKey);
+		ksDel (ks);
+		return EXIT_SUCCESS;
+	}
+
+	printf ("A real implementation would now \n");
+
+	Key * lookup = ksLookupByName (ks, BASE_KEY "/printversion", 0);
+	if (lookup != NULL && elektraStrCmp (keyString (lookup), "1") == 0)
+	{
+		printf ("print version information\n");
+
+		keyDel (errorKey);
+		ksDel (ks);
+
+		return EXIT_SUCCESS;
+	}
+
+	lookup = ksLookupByName (ks, BASE_KEY "", 0);
+	if (lookup != NULL && elektraStrCmp (keyString (lookup), "getter") == 0)
+	{
+		lookup = ksLookupByName (ks, BASE_KEY "/getter/keyname", 0);
+		if (lookup == NULL || strlen (keyString (lookup)) == 0)
+		{
+			printf ("report the error 'missing parameter: keyname'\n");
+			keyDel (errorKey);
+			ksDel (ks);
+
+			return EXIT_SUCCESS;
+		}
+		printf ("read the key '%s'\n", keyString (lookup));
+
+		lookup = ksLookupByName (ks, BASE_KEY "/getter/verbose", 0);
+		if (lookup != NULL && elektraStrCmp (keyString (lookup), "1") == 0)
+		{
+			printf ("print where the read key value comes from\n");
+		}
+	}
+	else if (lookup != NULL && elektraStrCmp (keyString (lookup), "setter") == 0)
+	{
+		const char * keyname;
+
+		lookup = ksLookupByName (ks, BASE_KEY "/setter/keyname", 0);
+		if (lookup == NULL || strlen (keyname = keyString (lookup)) == 0)
+		{
+			printf ("report the error 'missing parameter: keyname'\n");
+			keyDel (errorKey);
+			ksDel (ks);
+
+			return EXIT_SUCCESS;
+		}
+		lookup = ksLookupByName (ks, BASE_KEY "/setter/value", 0);
+		if (lookup == NULL)
+		{
+			printf ("report the error 'missing parameter: value'\n");
+			keyDel (errorKey);
+			ksDel (ks);
+
+			return EXIT_SUCCESS;
+		}
+		printf ("write the value '%s' to the key '%s'\n", keyString (lookup), keyname);
+
+		lookup = ksLookupByName (ks, BASE_KEY "/setter/verbose", 0);
+		if (lookup != NULL && elektraStrCmp (keyString (lookup), "1") == 0)
+		{
+			printf ("print where the key value is stored now\n");
```

Comment:
`bin/optsCommands set abc -v` is not coming here, what I am doing wrong?

And `bin/optsCommands set -v abc` fails with `ERROR: Expected at least 2 non-option arguments, but only got 1` which is confusing for me.

Ideally, the shell recorder would check if the example program accepts arguments as expected.

Reply:
Because of posixly evaluation (otherwise commands don't work), `bin/optsCommands set abc -v` means "set key `abc` to value `-v`".

In `bin/optsCommands set -v abc` you are missing the value for the `set` command. `bin/optsCommands set -v abc x` should work.

Reply:
> in/optsCommands set abc -v means "set key abc to value -v".

Ok, this should be written where you write that `posixly = 1` is implied. Because we are not doing exactly what POSIXLY_CORRECT does but actually something else: the original meaning is "If it is set, parsing stops as soon as a parameter is found that is not an option or an option argument." but for us it is: "If it is set, parsing stops as soon as a parameter is found that is not **a sub-command,** an option or an option argument."

Reply:
In a sense, we are doing what `POSIXLY_CORRECT` does, namely once we encounter a string in `argv` that doesn't start with `-`  we stop looking for options **for the current command**.

In other words, the behaviour is as if a POSXILY_CORRECT `getopt_long` was invoked separately for each sub-command. And everytime `getopt_long` returns, we either start another run, or collect into `args=indexed`/`args=remaining` depending on the next element of `argv`.

I will clarify this.